lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,351 @@
+        self.state = module.params['state']
+        self.force = module.params['force']
+        self.size = module.params['size']
+        self.type = module.params['type']
+        self.passphrase = module.params['passphrase']
+        self.comment = module.params['comment']
+        self.changed = False
+        self.check_mode = module.check_mode
+        self.privatekey = None
+        self.fingerprint = {}
+
+        if self.type in ('rsa', 'rsa1'):
+            self.size = 4096 if self.size is None else self.size
+            if self.size < 1024:
+                module.fail_json(msg='''For RSA keys, the minimum size is 1024 bits and the default is 4096 bits.
+                                     Attempting to use bit lengths under 1024 will cause the module to fail.''')
+
+        if self.type == 'dsa':
+            self.size = 1024 if self.size is None else self.size
+            if self.size != 1024:
+                module.fail_json(msg='''DSA keys must be exactly 1024 bits as specified by FIPS 186-2.''')
+
+        if self.type == 'ecdsa':
+            self.size = 256 if self.size is None else self.size
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg='''For ECDSA keys, size determines the key length by selecting from
+                                     one of three elliptic curve sizes: 256, 384 or 521 bits.
+                                     Attempting to use bit lengths other than these three values for
+                                     ECDSA keys will cause this module to fail.''')
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+            try:
+                tmpfile = None
+                try:
+                    if self.passphrase:
+                        with tempfile.NamedTemporaryFile(delete=False) as tf:
+                            tmpfile = tf.name
+                        os.chmod(tmpfile, stat.S_IRWXU)
+                        with codecs.open(tf.name, "w", encoding='utf-8') as tf:
+                            tf.write("#!/bin/sh\necho %s" % quote(self.passphrase))
+                        env = {"DISPLAY": "fake", "SSH_ASKPASS": str(tmpfile)}
+                    else:
+                        args.extend(['-N', ''])
+                        env = {}
+
+                    if self.comment:
+                        args.extend(['-C', self.comment])
+                    else:
+                        args.extend(['-C', ""])
+
+                    self.changed = True
+                    subprocess.call(args, env=env)
+                finally:
+                    if tmpfile is not None:
```

Comment:
You should use `module.add_cleanup_file()` (somewhere above) instead of manually trying to clean up.